HPCC-18142 Unbound DESDL based ESP services should report unbound state

mayx opened the pull request:
- Add a method to base and DESDL binding classes to indicate
  if the binding is dynamic binding
- Set the isDynamicBinding attribute in navigation data retrieved
- Show "Unbound DESDL Service" when there are no methods and the binding
  is dynamic.

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [x] I have tested my changes in multiple modern browsers
  - [x] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 5f3c9d5e1bc7fa6cbf4f6519a51fac0014ff709c
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 98.071 seconds |


Install hpccsystems-platform-community_6.4.1-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |96 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

mayx commented:
<img width="212" alt="screen shot 2017-08-17 at 1 14 24 pm" src="https://user-images.githubusercontent.com/907992/29424638-7a1a7392-834e-11e7-9227-50387728269b.png">

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: cbf184cc97bceabae34d53cdcf3420f0b9c0f127
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.677 seconds |


Install hpccsystems-platform-community_6.4.1-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |96 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

mayx commented:
@afishbeck please review. Please check the attached image above for the look of the text.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 24657c6c066c012f0dff8ac11011ad69aaff57bb
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.605 seconds |


Install hpccsystems-platform-community_6.4.1-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |96 |96 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |733 |733 |0 |
|test (thor) |627 |627 |0 |
|test (roxie) |760 |760 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented:
@mayx The place where dynamicesdl2 shows up in the image has always seemed wrong to me.  That is where the first service name is supposed to go if there is one.  That name is the ESP binding name I think.

Can we get rid of that?  And maybe put the informational text right there?  "[No Services]" or if it fits "[No Service Bindings]".

mayx commented:
@afishbeck Sorry didn't see your comments earlier. Actually "dynamicesdl2" here is the service name, not the binding name. I was using the same name for the esp service and binding. After changing the name of the service in environment.xml, now it looks as the image attached.

mayx commented:
<img width="533" alt="screen shot 2017-08-30 at 10 48 46 am" src="https://user-images.githubusercontent.com/907992/29878708-3e27d168-8d71-11e7-80db-82da9bbf29d4.png">

afishbeck reviewed (approved):
@mayx looks fine

richardkchapman commented on esp/services/esdl_svc_engine/esdl_binding.hpp:
```diff
@@ -375,6 +375,7 @@ class EsdlBindingImpl : public CHttpSoapBinding
 
     bool usesESDLDefinition(const char * name, int version);
     bool usesESDLDefinition(const char * id);
+    virtual bool isDynamicBinding() { return true; }
```
const and override modifiers would seem appropriate here

richardkchapman commented on esp/platform/espbinding.hpp:
```diff
@@ -106,7 +106,7 @@ class esp_http_decl CEspBinding : public CInterface,
     virtual bool qualifyServiceName(IEspContext & context, const char * servname, const char * methname, StringBuffer & servQName, StringBuffer * methQName){return false;}
     virtual IRpcRequestBinding *createReqBinding(IEspContext &context, IHttpMessage *request, const char *service, const char *method){return NULL;}
-
+    virtual bool isDynamicBinding() { return false; }
```
const would seem appropriate here

richardkchapman commented:
@mayx a couple of minor comments

mayx committed:
HPCC-18142 Unbound DESDL based ESP services should report unbound state

- Add a method to base and DESDL binding classes to indicate
  if the binding is dynamic binding
- Set the isDynamicBinding attribute in navigation data retrieved
- Show "Unbound DESDL Service" when there are no methods and the binding
  is dynamic.

Signed-off-by: mayx <yanrui.ma@lexisnexisrisk.com>
